### ISSUE TITLE
Construct export tool (a compiler target)

### ISSUE THREAD

**Issue by arekbulski** [kaitai-io/kaitai_struct#377]
After investigating the issue a bit more closely (and I admit that I am far from implementing it...), I came to a conclusion that the Construct->Kaitai tool should be implemented on Construct side, but the Kaitai->Construct export tool should be implemented on your side. I will need some assisstance with it, namely with the compiler since I dont speak Scala at all. I can provide you with examples, what the compiler should spew out, but I probably wont be able to write Scala code for it.

I foresee one (admittedly huge problem): Construct API is not stable (at least at the moment). This seems to be mostly affecting classes that are brand new, but there were also some changes to classes that existed before I took over the project. If it would take Kaitai a year to ship changes, thats kind of a problem. I suppose the users could use the newest version from GitHub (build the compiler from sources), so that would somewhat mitigate if not solve the issue. 

I attach the map of completion, to be edited later.
![29967916-b8aa5c24-8f21-11e7-891a-47bd8eb827a5](https://user-images.githubusercontent.com/5385838/37299646-2d83edb4-2624-11e8-8e24-59f91a3b7643.png)

**Comment by GreyCat** [comment]
I suspect that Construct target would be very different from what's usually generated, i.e. we'll need to match declarative constructions, not procedural statements. That's totally possible too, though.

https://github.com/kaitai-io/kaitai_struct/issues/253 is progressing smoothly, so chance are it would be much easier to do all the tests after it would be completed. Actually, it can be used right now, we just need to add relevant code to generate tests, Python's translator probably should be fine as is?

**Comment by arekbulski** [comment]
I will start working on both import/export tools then, and should deliver a first batch of translated examples within say a few days. Can I count on you to handle writing the compiler scala code?

To return the favor, I will get on top of some other work items that I assigned myself to and kind of had it stuck in the backlog, most importantly the Construct->Kaitai export tool which will be implemented entirely on Construct side. Actually its very similar to the compiler infrastructure, except the pyYAML module would be generating the output instead of a custom text-writer class.

**Comment by GreyCat** [comment]
> Can I count on you to handle writing the compiler scala code?

Sure :)

**Comment by arekbulski** [comment]
Would it be more convenient if structs were created kind of imperatively like this?
```py
s = []
s.append('fieldname' / Int8ub)
return Struct(*s)
```
instead of proper syntax?
```py
Struct(
'fieldname' / Int8ub,
)
```

**Comment by GreyCat** [comment]
Probably it won't matter, as this would be whole new `Compiler`, akin to `GraphVizCompiler`, which would not use existing templating mechanism (=does not need to fit into existing header-contents-footer workflow).

**Comment by arekbulski** [comment]
Could the construct-compiler (lets call it the export tool?) have a weekly release cycle?

**Comment by GreyCat** [comment]
I don't quite understand what you call "release cycle". At least in the near foreseeable future, "releasing" would probably remain relatively time-consuming and thus not very common task — mostly because it means lots of PR, writing announces, posting announces on the news sites, etc.

On the other hand, if you just want current builds, they're available right now, both for Debian & Windows. Would that qualify for "wait for 5-7 minutes after the commit" release cycle?

**Comment by arekbulski** [comment]
Yes that is exactly what I meant. So this solves the issue of Construct having unstable API.

**Comment by arekbulski** [comment]
Progress report on the import tool:
```
    def test_exportksy():
        d = Struct(
            "num1" / BytesInteger(4),
            "num2" / Int32ub,
            "num2" / Float32b,
            "data1" / Bytes(4),
            "data2" / GreedyBytes,
            "array2d" / Array(5, Array(5, BytesInteger(1)))
        )
        d.export_ksy()
--------------------------------------- Captured stdout call ----------------------------------------
meta:
  id: unnamed_schema
seq:
- id: num1
  type: u4be
- id: num2
  type: u4be
- id: num2
  type: f4be
- id: data1
  size: 4
- id: data2
  size-eos: true
- id: array2d
  repeat: expr
  repeat-expr: 5
  type: type_0
types:
  type_0:
    seq:
    - id: x
      repeat: expr
      repeat-expr: 5
      type: u1be

```

**Comment by arekbulski** [comment]
Switched from pyYAML to ruamelYAML, it fixed the order of keys.

**Comment by GreyCat** [comment]
If that's going to be just a byte array, just drop `type: xxx`, that would be raw byte array.

**Comment by arekbulski** [comment]
No, that would be Bytes analog, I am asking about Prefixed.
https://construct.readthedocs.io/en/latest/api/tunneling.html#construct.Prefixed

**Comment by GreyCat** [comment]
Um, then just use what you've proposed — it should work ;)

**Comment by arekbulski** [comment]
Do I get it right, that fields can have `doc` tag but entire struct cannot?

**Comment by GreyCat** [comment]
Typespec (that's probably what you call "entire struct") can have `doc` + `doc-ref`.

**Comment by arekbulski** [comment]
I was thinking of something on `meta` level but closest thing I found was `title`.
Typespec would I think be a nested struct, not outer-most struct.

**Comment by arekbulski** [comment]
Uhm... what would be the analog of Pass?
https://construct.readthedocs.io/en/latest/api/streaming.html#construct.Pass

**Comment by GreyCat** [comment]
Probably there won't be direct equivalent. Default case for switch is `_`. There is no "default case" for enums, as (1) they work pretty different from Construct implementation, i.e. then don't convert integers <-> strings, but integers <-> constants, (2) currently their implementation is language-dependent.

**Comment by arekbulski** [comment]
Would `size: 0` be a valid attributespec?

**Comment by GreyCat** [comment]
From KS point of view, definitely yes. From individual languages point of
view, I'm not sure that all of them allow zero-sized arrays, but probably
most do.

**Comment by arekbulski** [comment]
@GreyCat 
The docs say that AttributeSpec tags "must" come in specified order. Is that really a hard requirement or just a way of emphasizing the importance of a style guide?

**Comment by GreyCat** [comment]
"Style guide" is called *style* guide for a reason of enforcing certain style: where several different behaviors are technically possible, we suggest one to follow. If someone does not do what style guide suggests, it fails to comply with style guide, but the code could be still compilable. If someone does not do what [language reference](http://doc.kaitai.io/ksy_reference.html) dictates, most likely it will result in compilation error.

So, the short answer: tags inside attribute spec can come in any order. Compiler actually has no way to even know of that order, it just gets it as unordered map.

**Comment by arekbulski** [comment]
Alright, for the record this is what the style guide says:
"When specifying an attribute, one MUST use the following order of keys"

**Comment by GreyCat** [comment]
Exactly. That "MUST" is to be interpreted as "in order to make style-compliant ksy, you must do that".

**Comment by KOLANICH** [comment]
I guess the order should be user-controllable by passing 3 lists, one is for properties (in `seq`, `instances`, `enums`, `params`), one is for types, and one is for `meta`.
for example `["id", "-orig-id", "type", ...]` if we wan a cmply with the style guide.

**Comment by GreyCat** [comment]
I'm not sure that style compliance is worth pursuing, at least now. For example, PyYAML generates arrays-in-maps like that:

```yaml
seq:
- id: foo
  type: bar
```

while style guide suggests:

```yaml
seq:
  - id: foo
    type: bar
```

**Comment by KOLANICH** [comment]
It's not about style complience much, it's just inconvenient when the things go in another order or even in rwndom order from run to run (this will have nasty effect on diffs and testing). That's why the order is specified in style guide, I guess. I mean though we may have some troubles with complying with the style guide, we need have an order, any order convenient for user, just for convenience.


BTW, ruamel.YAML used now [allows the style prescribed by style guide](https://yaml.readthedocs.io/en/latest/detail.html#indentation-of-block-sequences)

**Comment by arekbulski** [comment]
I agree that non-deterministic order would be problematic for diffs. The tool exports keys in deterministic order, and also style-compliant as far as I seen it. This issue is considered solved.

I will fix the indentation, thanks to @KOLANICH .

**Comment by arekbulski** [comment]
What is the analog of Flag (boolean), u1 and b1?

**Comment by KOLANICH** [comment]
b1 is a single bit, u1 is a single byte.

According to https://github.com/construct/construct/blob/7692f818cf0757e0c554e4c4439f32ebdc5215e0/construct/core.py#L1554 it seems like u1.

**Comment by arekbulski** [comment]
Flag is actually bit/byte agnostic, so inside Bitwise it would translate to `b1` but you are right of course.

FlagsEnum, according to Construct semantics, skips unmmaped bits (so I would prefer them not having IDs). But I will consider it.

I am yet to implement BitStruct, would following be valid 2 byte data?
```yaml
- type: b4
- type: u1   <-- byte-unaligned
- type: b4
```

**Comment by KOLANICH** [comment]
There are some undesired byte-alignment issues, see https://github.com/kaitai-io/kaitai_struct_formats/pull/68 for more info.

**Comment by arekbulski** [comment]
Okay, I get it. 

I am not familiar with the import tag, could you craft me one for VarInt analog?
https://github.com/construct/construct/blob/master/construct/core.py#L1267-L1277
http://formats.kaitai.io/vlq_base128_le/index.html

**Comment by KOLANICH** [comment]
I don;t understand what you mean, but https://github.com/kaitai-io/kaitai_struct_formats/blob/master/serialization/google_protobuf.ksy#L7 may help

**Comment by arekbulski** [comment]
type: vlq_base128_le, got it.

Is there a Timestamp analog? 
https://construct.readthedocs.io/en/latest/api/misc.html#construct.Timestamp

**Comment by GreyCat** [comment]
As far as I understand, "Timestamp" does not parse anything, but rather converts existing integer into an object of a class of a certain library and vice versa. The simplest way to convert would be just to keep the subcon and that's it.

Probably one day we'll introduce something that would convert things to language- and library-specific implementations, but that's a major task to be undertaken.

Right now we have an initiative to support that at least in Web IDE: https://github.com/kaitai-io/kaitai_struct_webide/issues/51

**Comment by arekbulski** [comment]
Okay, I just made it provide the subcon (so its "type: u4" and just that).

Is there a way to (persistently) change stream offset, or read it?
https://construct.readthedocs.io/en/latest/api/streaming.html#construct.Seek
https://construct.readthedocs.io/en/latest/api/streaming.html#construct.Tell

**Comment by KOLANICH** [comment]
`_io.pos` should give a position, there is no seek in KS language (by design because seek is stateful), it is possible to use substreams and instances instead, but it'd be hard to replace seek.

**Comment by arekbulski** [comment]
You sure? I am asking about stateful seeking, not Pointer analog.
https://construct.readthedocs.io/en/latest/api/streaming.html#construct.Pointer

**Comment by KOLANICH** [comment]
there is no stateful seeking in the language by design.

**Comment by webbnh** [comment]
The runtime provides a seek method, e.g. in C++:
```
void
getNextRecord(kaitai::kstream& ks) {
    while (!ks.is_eof()) {
        iostream::pos_type curPos = ks.pos();

        try {
            rec = make_shared<myrecords_t::record_t>(&ks);
        } catch(...) {
            // Note that there is now an exceptional condition pending on the underlying istream, and any subsequent
            // access to the kstream (including is_eof()) is likely throw another exception.  Unfortunately there's no
            // KS method to clear it, and we don't have access to the istream, here...so just punt.
            break;
        }

        if (rec->start() != 0x01) {
            ks.seek(1 + curPos);
        }
}
```

**Comment by arekbulski** [comment]
Of course Python runtime also has such a method. The issue is, how do we call it using KSY?
https://github.com/kaitai-io/kaitai_struct_python_runtime/blob/master/kaitaistruct.py#L78-L79

**Comment by GreyCat** [comment]
As @KOLANICH said, by design, `seq` elements are in sequence, and thus can't be positioned arbitrarily in the stream. If you want elements positioned arbitrarily, then `instance` + `pos` is your friend, and it is indeed compiled into something that invokes `seek()`.

To convert something like "element1, element2, seek(x), element3, element4" from Construct, you'll have to:

* Render `element1` and `element2` into `seq` of a type (say, `type1`)
* Create a second type (say, `type2`), render `element3` and `element4` into its `seq`
* Add instance like that `type1`:

```yaml
instances:
  elements3_and_4:
    pos: x
    type: type2
```

**Comment by arekbulski** [comment]
Yeah, thank you @KOLANICH and yes I understand that @GreyCat . For the record, Construct also processes Struct members in sequence, and Structs are declarative, but doesnt mean fields cannot have (intended) side-effects. Seek is an example of that. Also you can have Pointer members and their order within a struct defines order in which they are parsed (their order of parsing is not defined by offsets). 

I suggest we allow AttributeSpec to have "pos", to allow stateful jumping? New feature?
```yaml
seq:
- id: field1
- size: 0    <-- so it doesnt consume bytes
  pos: offsetfield   <-- stateful stream seek
- id: field3
```

I will make Pointer exportable (using instances) and Seek just being unsupported.

**Comment by KOLANICH** [comment]
@arekbulski, could you provide as much stateful seeking use-case examples as possible? I mean short examples without anything unneeded and with some description what it does.

For the example you have provided it is a tear, these tears in seq are either filled aith empty dummy fields
```yaml
  seq:
    - id: field1
   - size: offsetfield #offset relative to end of field 1
   - id: field3
```
(take a note that there is no sizeof #84 so we have compute offsets manually)   

or we use instances.

In more complex situations like unions we use the following trick
```yaml
types:
  placeholder:
    seq:
      - size-eos: true
  some_union:
    seq:
      - id: placeholder
        size: 8
        type: placeholder
   instances:
     f:
       pos: 0
       io: placeholder._io
       type: f8
     u:
       pos: 0
       io: placeholder._io
       type: u8
```

**Comment by arekbulski** [comment]
Seek and Pointer are kinda interchangable. I dont think there is a use case where only one of these could be used. But if the user already used Seek then his schema wont be exportable. I am just trying to support as many classes as possible, just that. Which is a considerable issue I might add, most classes have no analog in Kaitai and Seek is one example of that.

**Comment by arekbulski** [comment]
Actually, strike my last. Seek and Pointer are not interchangable. Pointer reverts to last position (its non-stateful in that sense). Using Seek instead would require first using Tell to remember the offset, then a second Seek to jump back. Construct has way more building blocks than Kaitai, its as simple as that. I am just trying to support as many classes in the KSY exporter as possible.

**Comment by GreyCat** [comment]
> I suggest we allow AttributeSpec to have "pos", to allow stateful jumping? New feature?

No, you're missing the point here. That's not a "new feature", that's transforming declarative specification (i.e. describing *what* is in the structure) into imperative specification (i.e. describing *how* to read/write the structure).

By eliminating that "state" we're actually achieving lots of flexibility and freedom: i.e. when we have a sequence of `(a, b, c)`, a particular implementation might want to skip "a" and "b" and directly to "c". If "b" actually does something that affects parsing of "c" (i.e. changes some mode, reading position, encoding, etc), then this contract is broken and we don't have "attributes a, b, c", we have "instructions on how to read attributes a, b, c". With that, we can't read "c" until we read "a" and "b".

> Construct has way more building blocks than Kaitai, its as simple as that.

That's your opinion. In my opinion, I don't see a single situation where these "Seek" and "Tell" blocks can't be converted into primitives that KS offers (i.e. sequence attributes and parse instances).

**Comment by arekbulski** [comment]
In Construct, Struct members order does define order in which they are processed. If you put two Pointer fields, they will be processed in that order.
```py
Struct(
    'offset1' / Pointer(0xBEEF),
    'offset2' / Pointer(this.offset1),
    'data' / Bytes(this.offset2),
)
```

**Comment by GreyCat** [comment]
If I'm understand what it does properly, I believe this would work just fine, i.e. something like:

```yaml
seq:
  - id: data
    size: offset2
instances:
  offset1:
    pos: 0xbeef
    type: u4
  offset2:
    pos: offset1
    type: u4
```

Sequence parsing starts with `data`, which requests `offset2`, which requests `offset1`, which goes to 0xbeef, reads the integer, and rolls it all back the stack.

**Comment by GreyCat** [comment]
Actually, this *will* cause problems on the KS->Construct conversion, as I can imagine a certain pattern which won't be possible to put into Construct's `Struct(...)` sequence uniquely. For example:

```yaml
seq:
  - id: data
    size: 'switch_byte == 1 ? alt1 : alt2'
instances:
  switch_byte:
    type: u1
    pos: 0x1000
  alt1:
    type: u4
    pos: 0x2000
  alt2:
    type: u4
    pos: 0x3000
```

Size of `data` here depends on `switch_byte`, which obviously should be read before `data`, and after that, a decision should be made whether to read `alt1` or `alt2`, based on contents of `switch_byte`. Would it possible to do that in Construct?

**Comment by arekbulski** [comment]
Both alt1/alt2 would be parsed and inserted into context, regardless which one would be used, and would fail if parsing outside of available data (beyond EOF). Pointers are evaluated greedily. It might be possible to use IfThenElse so only one alt would be parsed.
```py
Struct(
    'switch1' / Pointer(0x3000, Int8ub),
    'alt1' / Pointer(0x1000, Int32ub),
    'alt2' / Pointer(0x2000, Int32ub),
    'data' / Bytes(lambda this: alt1 if this.switch1 == 1 else alt2),
)
```
I guess the bottom line is, Struct members would contain both "seq" and "instance" members, but since instances are lazily evaluated, there is no order between them (as far as KS sees it). If you translate those fields in the order defined by user, it will be fine, but kinda by chance.

**Comment by GreyCat** [comment]
> Both alt1/alt2 would be parsed and inserted into context, regardless which one would be used, and would fail if parsing outside of available data (beyond EOF).

That's not exactly equal, and that's not always possible. For example, one of these alternatives might result in infinitely recursive parsing (or at least a very long and resource-consuming process), if chosen improperly. This is actually frequently the case for parsing of filesystem images.

**Comment by arekbulski** [comment]
This better? The problem is, previous example is directly translatable. This one aint.
```py
Struct(
    'switch1' / Pointer(0x3000, Int8ub),
    'alt1' / If(this.switch1 == 1, 
        Pointer(0x1000, Int32ub), )
    'alt2' / If(this.switch1 != 1, 
        Pointer(0x2000, Int32ub), )
    'data' / Bytes(lambda this: alt1 if this.switch1 == 1 else alt2),
)
```

**Comment by arekbulski** [comment]
Actually I might have a solution. I might add a wrapper that lazifies arbitrary fields, using [lazy-object-proxy](https://github.com/ionelmc/python-lazy-object-proxy) which was found by @KOLANICH . Then the context would have both alt1/alt2 entries but those would be proxies.
```py
Struct(
    'switch1' / Pointer(0x3000, Int8ub),
    'alt1' / Lazy(Pointer(0x1000, Int32ub)),
    'alt2' / Lazy(Pointer(0x2000, Int32ub)),
    'data' / Bytes(lambda this: alt1 if this.switch1 == 1 else alt2),
)
```

**Comment by arekbulski** [comment]
What would be the bitwise analog to signed integers? Its not `b4` thats unsigned. And little endian?

**Comment by GreyCat** [comment]
> What would be the bitwise analog to signed integers?

There is no simple bit-level solution for signed integers, as technically sign can appear in tons of different ways — see https://github.com/kaitai-io/kaitai_struct/issues/121 for gory details. There's a proposal that (kind of) addresses it at https://github.com/kaitai-io/kaitai_struct/issues/342, but it's relatively low priority. Typical solution for now is declaring bit values + sign separately, and then use a value instance to do the calculation of the signed result.

> And little endian?

"Endianness" concept is also kind of very ambiguous for bit-level integers. However, lots of people rely on certain understanding of it, which can be implemented (see https://github.com/kaitai-io/kaitai_struct/issues/155#issuecomment-331413570), and would probably be supported in next versions.

**Comment by arekbulski** [comment]
Okies, I will just have signed and swapped bitwise integers as unsupported. 

Just for progress report, I worked a bit on the KSY exporter. Its almost finished.You can find all implemented and yet-to-implement classes on [this list.](https://github.com/construct/construct/blob/39738dfcb0343e3a40f874d374d77e66f819a7bc/tests/test_core.py#L1642)  Everything that aint on this list,  is not exportable.

**Comment by arekbulski** [comment]
Instances can be conditional? How does this work exactly?
```yaml
instances:
  test_if_b1:
    value: 123
    if: bits_a == false
```

**Comment by arekbulski** [comment]
Is this padded then terminated, or vice versa?
```yaml
  - id: str_term_and_pad
    size: 20
    terminator: 0x40
    pad-right: 0x2b
```

**Comment by arekbulski** [comment]
There is example using a::b::c syntax but I cannot find it in the documentation. Is it even in the documentation?

**Comment by GreyCat** [comment]
> Instances can be conditional? How does this work exactly?

Yes, they can. Take a look at the compiled code — it just returns `null` or nearest equivalent if the condition is not satisfied, and evaluation of value or parsing never takes place. `if` thus acts as a safeguard.

> Is this padded then terminated, or vice versa?

Terminator is always one byte and it follows exactly the string content (as in C). Padding follows that up to the size. I.e. `"abc"` encoded in that example you've mentioned would be:

    61 62 63 40|2b 2b 2b 2b|2b 2b 2b 2b|2b 2b 2b 2b|2b 2b 2b 2b 

> There is example using `a::b::c` syntax but I cannot find it in the documentation. Is it even in the documentation?

I was only implemented like several days ago. No, it's not yet in the documentation.

**Comment by arekbulski** [comment]
It took me a while to realize that NullTerminated supercedes NullStripped, so there is no point having both. NullStripped can be safely ommited. Correct?

**Comment by arekbulski** [comment]
I translated about half of the test formats into Construct. You can find them in this [branch](https://github.com/kaitai-io/kaitai_struct_tests/commits/translated-construct). The problem is, I am not entirely confident about some of those because I am gonna need to add at the very least 4 new classes into Construct. Since those do not exist yet, they dont have clear semantics, so I am kinda speculating here.

**Comment by arekbulski** [comment]
I am not sure if the compiler should insert `CString` or `NullTerminated(GreedyString)`. The first one supports UTF16/32 but the latter might be easier to (auto) translate.

**Comment by GreyCat** [comment]
@arekbulski Could you explain how exactly do you plan to use these files? They have .py extension, but right now running them would result in: (1) an error message because of lack of construct imports, (2) nothing resulting, as this is not a module exporting anything. We need some kind of wrapper for these to be able to at least run tests against them.

And, again, we'll need the basic idea of what test would look like. Ideally, could we reuse existing Python test specs with these files?

**Comment by arekbulski** [comment]
Oh these arent test cases, those are hand translated tags, so you know what the compiler/exporter should output for each tag. Isnt that what you asked earlier?

**Comment by arekbulski** [comment]
Yeah, but you said that you want to start working on the exporter and needed examples so thats why I wrote them. I will change them to include imports and export the topmost struct under some name like `main` and meta-id. Is there anything else?

**Comment by GreyCat** [comment]
You don't have to change it all manually on all the files, you can just show me on one file what would be the preferred output ;)

Also, it would be cool to show me how test specs should look like. I.e. what's you'd like to see in place of [test_hello_world.py](https://github.com/kaitai-io/kaitai_struct_tests/blob/master/spec/python/test_hello_world.py)?

**Comment by arekbulski** [comment]
I will update the examples but to not keep you waiting, template overall:
```py
from construct import *
from construct.lib import *
# if meta id was x
# topmost struct is always exported as _schema but also under name choses in meta ID
x = Struct(...)
_schema = x
```
I will get you some test cases ASAP.

**Comment by arekbulski** [comment]
Actually a test case would look like:
```py
from construct import *
from construct.lib import *
from helloworld import _schema
def test_schema():
  r = _schema.parse_file(filename)
  # pytest shows both sides of assertion, no need for assertEqual
  assert r.one == 80
```

**Comment by GreyCat** [comment]
The very first implementation of `-t construct` was just comitted and hopefully it would pass CI in several minutes. Result for `ksc -t construct hello_world.ksy` is:

```python
from construct import *
from construct.lib import *
hello_world = Struct(
	'one' / Int8ub,
)
_schema = hello_world
```

Unfortunately, I can't run the test now, looks like I've missing pytest or something...

**Comment by KOLANICH** [comment]
How about a test on self-consistency (ksy 1->construct->ksy 2 and kake sure they match (of course the files should be such that must match after the cycle))

**Comment by GreyCat** [comment]
Here you go, guys: https://github.com/kaitai-io/ci_targets/tree/master/compiled/construct — all formats get compiled into *something*.

Of course, there's lots of garbage there with `???` like this [term_strz.py](https://github.com/kaitai-io/ci_targets/blob/master/compiled/construct/term_strz.py) demonstrates.

**Comment by GreyCat** [comment]
> ksy 1->construct->ksy 2 and make sure they match

I don't think it would be that easy. There are several major differences between KS and Construct, thus KS -> Construct will always lose some information (at least instances vs seq), and Construct -> KS will have to invent more complex subtypes.

**Comment by KOLANICH** [comment]
>I don't think it would be that easy. There are several major differences between KS and Construct, thus KS -> Construct will always lose some information (at least instances vs seq), and Construct -> KS will have to invent more complex subtypes.

Yes, but after some cycles I expect convergence.

**Comment by arekbulski** [comment]
Michail is right, there are significant differences in infrastructure. Not only some details get lost in translation but also the exported KSY is messy, not quality code like you would write by hand. Compare [current output](https://gist.github.com/arekbulski/0036ee9c698a51ab2bb2275d394c3deb) and laugh/cry as much as you want.

**Comment by GreyCat** [comment]
Just pushed some strings support as well. Going to somewhat more serious matters, I'll ask some questions in comment to your work, @arekbulski.

**Comment by arekbulski** [comment]
Please dont use comments on commits. There is no way for me to navigate those other than Search page for your nickname. Better paste codes on this issue (or a new one) and ask.

**Comment by GreyCat** [comment]
Um, I believe you'll be getting an e-mail for all of them, but ok, I'll repost them here.

**Comment by arekbulski** [comment]
I might get an email but on the page, there is no listing. If the page is 100 screens long then its just impossible to find your comments, other than F3. Comments on PRs are easy to work with but comments on commits are problematic.

**Comment by arekbulski** [comment]
What would you like me to work on next?

**Comment by GreyCat** [comment]
@arekbulski Setting up test environment, something like `run-construct` in test repository that will run the tests specs with Construct against generated files? And a few test specs coded manually for starter would be cool ;)

**Comment by arekbulski** [comment]
FixedSized NullTerminated NullStripped added, in Tunneling (see tutorial page)
Those were used in examples I made earlier but were not implemented yet. To be shipped.
https://construct.readthedocs.io/en/latest/tunneling.html#working-with-bytes-subsets

There is also Lazy wrapper in the works, the one we talked about regarding instances/pointers.

**Comment by arekbulski** [comment]
There is a problem with order of types. The referred types need to exist in the script before where they are referred. Otherwise it goes kaput. 
https://github.com/kaitai-io/ci_targets/blob/master/compiled/construct/if_struct.py
NameError: name 'if_struct__arg_tuple' is not defined

**Comment by GreyCat** [comment]
> There is a problem with order of types.

This is actually bad. I can, of course, run toposort before dumping these, but ultimately it will fail on various self-referencing / recursive structures, like most filesystems are. Is it possible to, I don't know, declare some kind of a placeholder that we'll replace with the real `Struct` later? May be some magic with lambdas / lazy wrapper again?

**Comment by GreyCat** [comment]
I've managed to run the test spec you've proposed, by replacing pytest with our existing unittest (there's actually good idea to do that, as I don't want to go with all that XML reporting integration again with a new framework). So far it fails with:

```
    r = _schema.parse_file(filename)
AttributeError: 'Struct' object has no attribute 'parse_file'
```

What am I missing?

**Comment by GreyCat** [comment]
Ok, I've figured it out: I just needed a newer Construct that has `.parse_file(...)`. The one shipped in Debian (2.8.16-0.2) does not support it. So, here we go, we have a `run-construct` in the tests now.

**Comment by GreyCat** [comment]
I've added basic test automation for `construct2` and `construct3` targets — and it works —
 see http://kaitai.io/ci/ — although only one test is implemented now — "hello_world".

Congratulations, Construct target is now officially first-class citizen in KS :) Feel free to port some tests manually or just wait until I'll get it running using KST...

**Comment by GreyCat** [comment]
Just committed basic Construct specs generation to KST, and, boom, we have 100 specs ported to test Construct target already ;)

Out of 100 tests, 11 result in "failure", and 68 result in "error" => 21 tests pass! Not bad for a start ;)

**Comment by arekbulski** [comment]
Wow let me look for that champaign bottle I had somewhere...

The parse_file method was added only recently (.39) and NullTerminated/NullStripped is yet to be shipped, so we will need literally most recent version on pypi (or upstream from github).

To solve the circular reference issue, we could use LazyBound. Look at the first example in docstring. I think the cleanest code could be produced by toposorting it and LazyBound-ing selectively some structs to break the circularity, but keeping it to minimum. In schemes that have no circularity, there would be no LazyBound. On the other hand, we could just wrap every struct in it. It would generate valid code, just not very clean one.
https://construct.readthedocs.io/en/latest/api/lazy.html#construct.LazyBound

**Comment by GreyCat** [comment]
Currently Travis integration install latest pip version, that is 2.9.39.

However, it seems that lots of test fail due to lack of http://construct.readthedocs.io/en/latest/api/tunneling.html#construct.FixedSized — that's some new addition as well?

**Comment by arekbulski** [comment]
Yes, it was added like yesterday. I just shipped version .40 to pypi. FixedSized included.

**Comment by GreyCat** [comment]
* Added `if` support
* Added `pos` support -> Pointer
* Properly set up `nowClass` for TypeProvider => fixed compilation of some expressions
* Added LazyBound type resolution everywhere (that's easy to add and it seems to work -> may be we will do better toposort and/or inlining in future)

New result: 27%

**Comment by arekbulski** [comment]
I recommend `-construct-render: inlineStruct` for inlining structs?

**Comment by arekbulski** [comment]
I looked at the other half of formats and honestly I dont know how to translate them, I dont even understand  how some of those work. I will need your help with further translations. Please tell me which ones you need and I will try to get those done, or ask you for assisstance.

Meanwhile I have some thoughts to share, about the basics. There are alternatives to consider:

SIZE: N without type
- Bytes(N) seems good default for parsing, but if user has intentions to build then its something like 50/50 that its either Bytes or Padding. This class needs a build value of equal length.
- Padding(N) builds without value. I recommend `-construct-render: Padding`.

SIZE: N, TYPE: X
- FixedSized(N, X) is a safe default, it always works but it uses substreams
- Padded(N, X) uses stream tell which has better performance especilly if N is in megabytes. I recommend `-construct-render: Padded`

Which leads me to another point that I noticed, there could be >1 render option, although it would not be that difficult to split this into 2 fields (Flag+Padding) within the KSY:
```yaml
- size: 10
  type: u1
  -construct-render: Padded,Flag
```

**Comment by GreyCat** [comment]
@arekbulski I don't think we should ever think of any optimizations before majority of the test cases (say, 80-90%) would work. Let's keep things as simple as possible until we'll get it working. I encourage you to just take a look at the test one-by-one and tell me what's wrong with each of them.

You can actually click on each square in [CI](http://kaitai.io/ci/) and see the relevant test run messages and fail stack traces. For example, for `BcdUserTypeBe`:

<details>
Traceback (most recent call last):
  File "/home/travis/build/kaitai-io/kaitai_struct/tests/spec/construct/test_bcd_user_type_be.py", line 9, in test_bcd_user_type_be
    r = _schema.parse_file('src/bcd_user_type_be.bin')
  File "/home/travis/.local/lib/python2.7/site-packages/construct/core.py", line 305, in parse_file
    return self.parse_stream(f, **contextkw)
  File "/home/travis/.local/lib/python2.7/site-packages/construct/core.py", line 296, in parse_stream
    return self._parsereport(stream, context, "(parsing)")
  File "/home/travis/.local/lib/python2.7/site-packages/construct/core.py", line 308, in _parsereport
    obj = self._parse(stream, context, path)
  File "/home/travis/.local/lib/python2.7/site-packages/construct/core.py", line 2067, in _parse
    subobj = sc._parsereport(stream, context, path)
  File "/home/travis/.local/lib/python2.7/site-packages/construct/core.py", line 308, in _parsereport
    obj = self._parse(stream, context, path)
  File "/home/travis/.local/lib/python2.7/site-packages/construct/core.py", line 2567, in _parse
    return self.subcon._parsereport(stream, context, path)
  File "/home/travis/.local/lib/python2.7/site-packages/construct/core.py", line 308, in _parsereport
    obj = self._parse(stream, context, path)
  File "/home/travis/.local/lib/python2.7/site-packages/construct/core.py", line 4683, in _parse
    return self.subcon._parse(io.BytesIO(data), context, path)
  File "/home/travis/.local/lib/python2.7/site-packages/construct/core.py", line 5439, in _parse
    return sc._parsereport(stream, context, path)
  File "/home/travis/.local/lib/python2.7/site-packages/construct/core.py", line 308, in _parsereport
    obj = self._parse(stream, context, path)
  File "/home/travis/.local/lib/python2.7/site-packages/construct/core.py", line 2067, in _parse
    subobj = sc._parsereport(stream, context, path)
  File "/home/travis/.local/lib/python2.7/site-packages/construct/core.py", line 308, in _parsereport
    obj = self._parse(stream, context, path)
  File "/home/travis/.local/lib/python2.7/site-packages/construct/core.py", line 2567, in _parse
    return self.subcon._parsereport(stream, context, path)
  File "/home/travis/.local/lib/python2.7/site-packages/construct/core.py", line 308, in _parsereport
    obj = self._parse(stream, context, path)
  File "/home/travis/.local/lib/python2.7/site-packages/construct/core.py", line 2703, in _parse
    return self.func(context) if callable(self.func) else self.func
  File "/home/travis/.local/lib/python2.7/site-packages/construct/expr.py", line 160, in __call__
    lhs = self.lhs(obj) if callable(self.lhs) else self.lhs
  File "/home/travis/.local/lib/python2.7/site-packages/construct/expr.py", line 160, in __call__
    lhs = self.lhs(obj) if callable(self.lhs) else self.lhs
  File "/home/travis/.local/lib/python2.7/site-packages/construct/expr.py", line 160, in __call__
    lhs = self.lhs(obj) if callable(self.lhs) else self.lhs
  File "/home/travis/.local/lib/python2.7/site-packages/construct/expr.py", line 160, in __call__
    lhs = self.lhs(obj) if callable(self.lhs) else self.lhs
  File "/home/travis/.local/lib/python2.7/site-packages/construct/expr.py", line 160, in __call__
    lhs = self.lhs(obj) if callable(self.lhs) else self.lhs
  File "/home/travis/.local/lib/python2.7/site-packages/construct/expr.py", line 160, in __call__
    lhs = self.lhs(obj) if callable(self.lhs) else self.lhs
  File "/home/travis/.local/lib/python2.7/site-packages/construct/expr.py", line 160, in __call__
    lhs = self.lhs(obj) if callable(self.lhs) else self.lhs
  File "/home/travis/.local/lib/python2.7/site-packages/construct/expr.py", line 160, in __call__
    lhs = self.lhs(obj) if callable(self.lhs) else self.lhs
  File "/home/travis/.local/lib/python2.7/site-packages/construct/expr.py", line 188, in __call__
    return self.__parent(obj)[self.__field]
KeyError: 'digit8'
</details>

From a broader perspective, I see a few major areas that need attention:

* Enums -> I reckon that enum literals (i.e. stuff like `enum_type::element`) should be translated into just strings for Construct?
* BytesPadTerm/StrPadTerm - needs to check what should be fixed there to parse stuff properly
* All `Switch*` stuff - you need to tell me how to render switching types
* `process: XXX` tag is just ignored now and thus all `Process*` tests fail; I reckon there was some way to use `Transform()` in Construct to do these?
* Navigation: `_root` and `_parent` - can you tell me how should I translate these for Construct?
* `repeat-until` is left untranslated now as well
* Expressions:
  * FloatToI shows that one can't just do `Computed(int(this.single_value))` with `'single_value' / Float32l` — it fails with mysterious `TypeError: __call__() takes at least 2 arguments (1 given)`
  * Access to `_io` and, consequently, `_io.eof` and `_io.pos` in expressions — right now they are translated into `this._io.is_eof()` and thus just fail with the same `__call__() takes at least 2 arguments`.
  * Doing `len(this.afloat)` to get a number of elements in array doesn't seem to work — see ExprArray test — `TypeError: object of type 'Path' has no len()`
* Bit-level integers are relatively complex, as they would require application of some heuristics and make compiler track modality of current parsing (i.e. normal `Struct` or `BitStruct`).

**Comment by arekbulski** [comment]
Enum values can be translated into string literals, yes. 
Getting on top of Switch stuff, sir! 

Padded and Terminated are not correctly translated, compare:
https://github.com/kaitai-io/ci_targets/blob/master/compiled/construct/str_pad_term.py
https://github.com/kaitai-io/kaitai_struct_tests/blob/translated-construct/_construct/str_pad_term.py
https://github.com/kaitai-io/ci_targets/blob/master/compiled/construct/bytes_pad_term.py
https://github.com/kaitai-io/kaitai_struct_tests/blob/translated-construct/_construct/bytes_pad_term.py

The `_parent` translates literally as `_` like in `this._.field`. Others do not exist yet, please use these in the meantime while I implement those:
- `_root` as `_root` (a dict)
- `_io` as `_stream` (a BytesIO)
- `_io.pos` as `_streampos()` (an integer function)
- `_io.size` as `_streamsize()` (an integer function)
- `_io.eof` as `_streameof()`(a boolean function)

**Comment by GreyCat** [comment]
Ok, I've fixed:

* `_parent`, `_root`, `_io`
* `terminator` and `pad-right` translation

This pumps up progress up to 32%.

I don't quite understand what you mean by "an integer function". How would you translate something like `_parent.foo.bar._io.pos`, for example?

* `this._.foo.bar._streampos()`?
* or `_streampos(this._.foo.bar)`?
* or `this._.foo.bar._stream._streampos()`?

**Comment by arekbulski** [comment]
Hmm, the `_parent.foo.bar._io.pos` I think should translate into `this._.foo.bar._streampos()` but... I dont understand how it would be different than `this._streampos()`. I assumed that _streampos() would be a context entry that is a lambda (not an integer value) that returns current offset at the time it gets evaluated. Therefore `this._streampos` and `this.foo.bar._streampos` is just the same function. Do I misunderstand the meaning of `_io.pos`?

--------------------------------
There is also some small issue with _root. You see, in Construct there is a "top context level" that is managed directly by the `parse/build` methods that essentially contains only the keywords that were passed by the user and few special stuff like `_parsing=True _building=False`. The topmost Struct, like every struct, does context nesting so that level becomes one up. Example seems in order:
```py
Computed(this.a).parse(a=1)
Struct(Computed(this._.a)).parse(a=1)
```
Those 2 lines of code should explain it: [Construct.parse](https://github.com/construct/construct/blob/4f4f851108ff5856a9249b14d39f2bbc3821f026/construct/core.py#L294) and [Struct._parse](https://github.com/construct/construct/blob/4f4f851108ff5856a9249b14d39f2bbc3821f026/construct/core.py#L2069)

How about we use 2 different entries? `_topmost` would be the top Struct context and `_root` would be the top level managed by parse method. Its kinda just a matter of naming things.

**Comment by arekbulski** [comment]
Regarding BitStruct: I think we could consider following approach. 
- default to Struct and group bN fields into nested structs (under some automatic id)
- if `-construct-render: BitStruct` then make it BitStruct and wrap non-bN fields in [Bytewise](https://construct.readthedocs.io/en/latest/api/bytes.html#construct.Bytewise) 

This approach seems problematic for end users, but it also makes it Zen. I dont remember the quote but paraphrasing: explicit over guessing. There are good reasons to not make it BitStruct by default. Bytewise is not compatible with every possible construct. Recently someone filed a [bug report ](https://github.com/construct/construct/issues/697)because GreedyRange and GreedyBytes (by extension GreedyString) are not working inside Bytewise. Partial fix is underway but in principle, not everything can work inside Bytewise.

**Comment by arekbulski** [comment]
In format `size: 5, repeat: until` is that 5 bytes total or 5 bytes per item?

**Comment by arekbulski** [comment]
I assume that if switch-on parses a value that has no mapping, it fails?

**Comment by arekbulski** [comment]
Translated the switch and repeatuntil related formats: https://github.com/kaitai-io/kaitai_struct_tests/commit/0d43297a70fa2ed5e484832495d0c37562760c48

**Comment by arekbulski** [comment]
How is this not a valid ksy?
```yaml
meta:
  id: demo
seq:
- id: x1
  size: 10
  type: u1
instances:
  lazy:
    value: _io.pos + x1._io.pos
```
$ ksc -t python demo.ksy 
/seq/0/size: unknown key found, expected: consume, doc, doc-ref, eos-error, id, if, include, repeat, terminator, type

**Comment by GreyCat** [comment]
> I assumed that _streampos() would be a context entry that is a lambda (not an integer value) that returns current offset at the time it gets evaluated. Therefore `this._streampos` and `this.foo.bar._streampos` is just the same function. Do I misunderstand the meaning of `_io.pos`?

`_io` gets that particular element's stream, not just "current stream". Consider this example for a 10-byte input stream:

```yaml
seq:
  - id: buf
    type: my_buf
    size: 5 # a limited substream of exactly 5 bytes
instances:
  my_size:
    value: _io.size # returns 10, as original stream is 10 bytes long
types:
  my_buf:
    instances:
      my_size_in_buf:
        value: _io.size # returns 5
      parent_size_in_buf:
        value: _parent._io.size # references parent's stream, and thus returns 10 again
```

> How about we use 2 different entries? `_topmost` would be the top Struct context and `_root` would be the top level managed by parse method. Its kinda just a matter of naming things.

I don't think I properly understand the difference. Could you explain to me what should it map to in any other language, or what would be the use case for difference between "top Struct context" vs "top managed by parse method" inside a ksy file?

> I assume that if switch-on parses a value that has no mapping, it fails?

No, it returns `null`.

> How is this not a valid ksy?

There's been a long discussion about that: https://github.com/kaitai-io/kaitai_struct_compiler/issues/77

In a nutshell, it's simple: in current implementation, `_io` is an attribute of KaitaiStruct (+a relevant getter method like `_io()`), and thus it is only available on user types. The initial (and current) rationale was that  we won't be able to store that substream anywhere and thus we've banned usage of `size` or `terminator` on a primitive types like integer. Strings and byte arrays allow `size` and `terminator`, but treat them slightly differently, so you still can't do `some_string._io`.

In some future, we'll probably fix that to act as per [this blueprint](https://github.com/kaitai-io/kaitai_struct_compiler/issues/77#issuecomment-370770274). However, it's relatively low-priority task.

**Comment by GreyCat** [comment]
`repeat-until` is implemented, as per your examples. Progress is 34%.

**Comment by arekbulski** [comment]
> _io gets that particular element's stream, not just "current stream".

Thank you for explaining. Construct current implementation does not expose streams or substreams. I will get back to you when I implemented some changes. For that matter, it currently does not allow selecting streams either so following is untranslatable:

```yaml
    instances:
      body:
        io: _root._io
        pos: ofs_body
        size: len_body
      something_in_big:
        io: _root.big_container._io
        pos: ofs_in_big
        size: len_in_big
```

> I don't think I properly understand the difference. Could you explain to me 

Sure. I wrote this example for you. I think I already told you that, user can externally provide values to be inserted into the context. That is what I consider _root level. The topmost struct (the seq that is on same level as meta) would do context nesting, so its one level lower than _root. 

Kaitai `_root` should be translated into Construct `_topmost`, and `_parent` into `_`.

```py
d =  # _root context level
    Struct( # _topmost context level
        Struct(
            Struct(
                Computed(_root.user),
                Computed(_topmost._.user), # notice up-level
                Computed(_._._.user),
            )
        )
    )
d.parse(user=1)
```

**Comment by arekbulski** [comment]
I translated few examples of instances: https://github.com/kaitai-io/kaitai_struct_tests/commit/85e560aedb00ab7c91a10057bfa50a7a4579355b

**Comment by arekbulski** [comment]
From what I gathered when doing translations, `process` tag can be used with both `size` and `size-eos`?

**Comment by GreyCat** [comment]
> From what I gathered when doing translations, `process` tag can be used with both `size` and `size-eos`?

Yes, it actually needs any kind of size determination - i.e. `size`, `size-eos`, or `terminator`.

**Comment by GreyCat** [comment]
> Sure. I wrote this example for you. I think I already told you that, user can externally provide values to
> be inserted into the context. That is what I consider _root level. The topmost struct (the seq that is on
> same level as meta) would do context nesting, so its one level lower than _root. 
>
> Kaitai `_root` should be translated into Construct `_topmost`, and `_parent` into `_`.

Ok, so it is related to user-passed parameters (`params` in KS) then. Probably equivalent in the KS would be something like that:

```yaml
meta:
  id: level1
params:
  - id: user
    type: u1
seq:
  - type: level2
types:
  level2:
    seq:
      - type: level3
  level3:
    instances:
      access_via_root:
        value: _root.user
      access_via_parent:
        value: _parent._parent.user # just two levels up, not three
```

So, in KS, there is no extra level of separation between user-passed parameters and normal sequence / instance attributes. By the way, if you have that separate layer for parameters, how do you pass parameters into inner subcons (even when they are declared as a distinct named variable)?

**Comment by arekbulski** [comment]
> By the way, if you have that separate layer for parameters, how do you pass parameters into inner subcons (even when they are declared as a distinct named variable)?

I dont understand the question...

**Comment by GreyCat** [comment]
> I dont understand the question...

How would you translate this?

```yaml
meta:
  id: params_call_short
seq:
  - id: buf1
    type: my_str1(5)
types:
  my_str1:
    params:
      - id: len
        type: u4
    seq:
      - id: body
        type: str
        size: len
        encoding: UTF-8
```

I reckon it would be something like

```python
my_str1 = Struct(
  'body' / Bytes(_.len)
)

params_call_short = Struct(
  'buf1' / my_str1 # how do I pass 5 here?
)
```

**Comment by GreyCat** [comment]
I've just added lazy instances rendering, and it fails with 

```
  File "kaitai_struct/tests/compiled/construct/expr_2.py", line 8, in 
    'len_mod' / Lazy(Computed(lambda this: (this.len_orig - 3))),
NameError: name 'Lazy' is not defined
```

Is that `Lazy` something new / unreleased, or am I doing things wrong?

**Comment by arekbulski** [comment]
I read the documentation regarding params and I think you and I have different ideas what it is. This example does not have externally provided parameter, its a parameterized constant internal to the schema. Externally provided means known at runtime (when parse method is called). 

I think this could (or needs to) be translated by concretising it. Like a template, not generics.
```py
my_str1_5 = Struct(
  'body' / Bytes(5)
)
params_call_short = Struct(
  'buf1' / my_str1_5,
)
```
or
```py
my_str1_5 = Struct(
  'len' / Computed(5),
  'body' / Bytes(this.len),
)
params_call_short = Struct(
  'buf1' / my_str1_5,
)
```

**Comment by GreyCat** [comment]
That is not always possible, or, at least, it is not always practical. For example, this value can be both stated inline (every time a different value) or passed from a totally different source, only available in run time:

```yaml
  - id: buf1
    type: my_str1(5)
  - id: buf2
    type: my_str1(10)
  - id: buf3
    type: my_str1(some_len)
```

This is how various library types work, like [bcd](http://formats.kaitai.io/bcd/index.html), for example. While it is technically probably possible to repeat that monstrosity in the output every time it is invoked, the point of parameters is actually to not do that, just declaring it once and reusing it again and again.

**Comment by arekbulski** [comment]
Okie dokie. Let me sit on this a while and I will think of some implementation and get back to you. I understand the idea, just not sure how to implement it yet.

I translated some expressions (inlcuding _io related) and am working on nav examples (_root and whatnot). Process tags also underway. Params on hold. What else would you like to me work on next? :)

**Comment by GreyCat** [comment]
I'll implement switches soon, and that would hopefully give us 60-70% coverage, which is good. We'll see then. Please take a look at [my problem with Lazy](https://github.com/kaitai-io/kaitai_struct/issues/377#issuecomment-376446919) here?

**Comment by arekbulski** [comment]
Sorry, I somehow didnt notice that one post earlier. Lazy is yet to be implemented. Also _stream is yet to be implemented. I will give those a priority.

**Comment by arekbulski** [comment]
Would it be a problem to have params attributes translate to both _root and _topmost anchored depending on details? I could have the topmost Struct copy over the entries from the root. Just that it would not be very clean and safe. But as Zen teaches, practicality over purity.

**Comment by GreyCat** [comment]
Meanwhile, I've implemented the basic switches, and this brings us to 42%.

**Comment by arekbulski** [comment]
Some progress on Switches, thats nice. 

I also made some changes to Construct today, KSY related:
- added Lazy, although I have to admit that lazy-proxy-object its a funky thing. It doesnt work exactly the way you might imagine it would. If you find it problematic to work with then tell me. Maybe I can reimplement it.
- added ProcessXor, examples will follow shortly.
- Struct exposes some new context enries.

Particularly:
- _stream is BytesIO with additional attributes (size eof seek tell) methods.
- _topmost (Kaitai _root)
- _root which, unless you want to meddle with params, lets skip this one

**Comment by GreyCat** [comment]
If you're *adding* `_stream`, `_root` and `_topmost` for sake of KS interoperability, why don't we keep the names, i.e `_io` => `_io`, `_root` => `_root`, and invent a new name for "root structure to access parameters', just to avoid confusion? If these existed long before, of course, that's out of question.

**Comment by arekbulski** [comment]
Well, sure why not. `_io _root` are fine, what would you name the other root?

**Comment by GreyCat** [comment]
No idea, really %) If it's sole purpose is accessing parameters, then may be `_params` would be better, so we could use `_params.foo` to indicate user-supplied parameter `foo`?

**Comment by arekbulski** [comment]
Actually I was thinking af exactly the same thing, but after a beer, hugging the pillow seemed more important than writing  a reply. :sleeping: 

`_io _root _params` it is.

**Comment by GreyCat** [comment]
JFYI, after last updates, all Construct tests fail with something like

```
Traceback (most recent call last):
  File "/home/travis/build/kaitai-io/kaitai_struct/tests/spec/construct/test_cast_to_imported.py", line 9, in test_cast_to_imported
    r = _schema.parse_file('src/fixed_struct.bin')
  File "/home/travis/.local/lib/python2.7/site-packages/construct/core.py", line 328, in parse_file
    return self.parse_stream(f, **contextkw)
  File "/home/travis/.local/lib/python2.7/site-packages/construct/core.py", line 319, in parse_stream
    return self._parsereport(stream, context, "(parsing)")
  File "/home/travis/.local/lib/python2.7/site-packages/construct/core.py", line 331, in _parsereport
    obj = self._parse(stream, context, path)
  File "/home/travis/.local/lib/python2.7/site-packages/construct/core.py", line 2081, in _parse
    obj = Container(_stream = FeaturedBytesIO(stream))
  File "/home/travis/.local/lib/python2.7/site-packages/construct/core.py", line 141, in FeaturedBytesIO
    stream.size = lambda: _size_stream(stream)
AttributeError: 'file' object has no attribute 'size'
```

**Comment by arekbulski** [comment]
The FeaturedBytesIO implementation is kinda messy. Not every stream has `size` attribute. You can however safely use `_size_stream` on every stream. Would it be problematic for you to change `_io.size` syntax into `_size_stream(_io)` so I could drop that class from Construct?

I could not find the exact commit that introduced this issue but I am pretty sure thats what it is.

**Comment by arekbulski** [comment]
Oh wait, I will try to fix this issue somewhat differently, by making all streams FeaturedBytesIO. But its gonna take some time.

**Comment by arekbulski** [comment]
This is the example that failed. Its referring to undefined helloworld?
https://github.com/kaitai-io/ci_targets/blob/master/compiled/construct/cast_to_imported.py

**Comment by arekbulski** [comment]
I looked into it and cant find what the problem is. Could you try changing the `_stream.size` attribute to `_size_stream(_stream)` instead? Not every stream is FeaturedBytesIO, the implementatin is sloppy.

**Comment by arekbulski** [comment]
Yeah I just confirmed what the problem is. On Python 2, a file stream cannot be extended with additional attributes. Meaning you need to wrap _io with 
- _seek_stream _tell_stream  (or use _io.seek _io.tell, the original methods)
- _size_stream _iseof_stream

Its not possible for it to work otherwise.

**Comment by arekbulski** [comment]
Problem was fixed (I will have the fix ship to pypi within hours). You need to wrap the `_io` stream with the following functions. Its not possible to add new attributes to file streams (on Python 2):
```py
def stream_seek(stream, offset, whence=0):
    try:
        return stream.seek(offset, whence)
    except Exception:
        raise StreamError("stream.seek() failed, offset %s, whence %s" % (offset, whence,))


def stream_tell(stream):
    try:
        return stream.tell()
    except Exception:
        raise StreamError("stream.tell() failed")


def stream_size(stream):
    fallback = stream.tell()
    end = stream.seek(0, 2)
    stream.seek(fallback)
    return end


def stream_iseof(stream):
    fallback = stream.tell()
    data = stream.read(1)
    stream.seek(fallback)
    return not data

```

**Comment by GreyCat** [comment]
Ok, understood, thanks! Will apply the fix now... We really need to get `construct3` and `python3` fixed on CI...

**Comment by arekbulski** [comment]
Just shipped the aforementioned fixes to pypi.

**Comment by GreyCat** [comment]
Applied the fix, we seem to be at 44% now.

**Comment by GreyCat** [comment]
> NavRoot

Seems to be fixed now.

> NavParentOverride

Um, no, that's the whole point of "parent override", that you can keep having same `_parent` expression, but in the generated instancing we can pass not a normal `self` as parent, but something else in some cases. I.e. for Python, normal invocation without override would be:

```python
self.child_1 = self._root.Child(self._io, self, self._root)
```

and with parent override:

```python
self.child_2 = self._root.Child(self._io, self._parent, self._root)
```

Note that 2nd argument (and that is parent we pass into the user type) is different than normal `self`.

If that's hard to implement, I'd suggest to just skip this whole "parent override" idea for Construct target, it's not critical by any means.

**Comment by arekbulski** [comment]
Uh, could you simply add a `"_"/Computed(...)` field that would override the context entry?

**Comment by GreyCat** [comment]
Ah, ok, probably I've understood what you mean here... Let's try it...

**Comment by GreyCat** [comment]
It would probably even work in this particular case, but the main problem is actually that this `Computed(...)` needs to be evaluated in context of parent, on in context of child where we'll be adding that field to.

**Comment by arekbulski** [comment]
I will get you Process XOR/rotate and few other formats translated in short time. From what I understand the CI page, the Enums and Switches are mostly red because KST doesnt generate the test cases?

**Comment by GreyCat** [comment]
> the Enums and Switches are mostly red because KST doesnt generate the test cases?

Right. Actually, here's the new coverage chart:

![coverage_construct2](https://user-images.githubusercontent.com/157154/38455241-3b627b64-3a6d-11e8-8fe3-19726c3ef877.png)

**Comment by arekbulski** [comment]
How do you feel about the two of use working on Construct today? I will try not to be so slopy.

**Comment by GreyCat** [comment]
Unfortunately, I only appear here from time to time today and can do only some small pieces of work.

Next big thing on my schedule is probably (1) porting the rest of switch specs to KST, (2) implementing nested type accessing for enum literals, and then (3) porting enums to KST. That would probably give us ~80-90% coverage for KST tests, and that should make adoption of any new emerging targets (Construct, Go, Rust, may be C) much easier.

Feel free to suggest whatever else I can fix in Construct generation. There are plenty of "red" things on that chart above that need fixing. Probably stuff like `expr_*`, `float_to_i`, `process_*` should be relatively easy to fix.

**Comment by arekbulski** [comment]
Okie dokie, I was just asking for someone to play with me with my toys. :smile: 

I pushed a new commit with translated formats: https://github.com/kaitai-io/kaitai_struct_tests/commit/a49e69738c9cd5c279a8229f1f79e9c37cc40994.
On Construct side: I need to add ProcessRotateLeft class, in Pointer add stream parameter (so you select a stream), and in Struct have _index copied over so you can refer to it directly. I will let you know immediately when that is implemented and shipped to pypi.

**Comment by arekbulski** [comment]
I kinda forsee a problem with `io` tag support (selecting stream). Can this feature be only used with pointer instances (those with `pos`) or with any tag, including seq?

**Comment by GreyCat** [comment]
> I kinda forsee a problem with `io` tag support (selecting stream). Can this feature be only used with pointer instances (those with `pos`) or with any tag, including seq?

Only with instances with `pos`. Sequence attributes are indeed dubbed "sequence" because they define data in sequence in the same stream.

This one is relatively important, as stream selection is used by a lot of production format specs. 26 out of 110 ksy files in formats repo use it.

> The ParentOverride is also a problem that I am unable to solve at this time.

Don't bother, really. This functionality is largerly superseeded by passing arbitrary parameters, so I don't think that anyone uses it in reality.

**Comment by arekbulski** [comment]
Okie dokie, then we are at actionable consensus. I will add `streamfunc` parameter to `Pointer` class.

**Comment by GreyCat** [comment]
@arekbulski, @KOLANICH — is there any chance that you'd be still interested in this issue?

I've finally completed enum support in KST, this can be leveraged to add another huge chunk to construct target support.

**Comment by arekbulski** [comment]
Construct is no longer actively developed. I am planning to resume workin on it for a few days, finish what was left unfinished and then freeze it. So no atm and yes possibly in the future.

**Comment by arekbulski** [comment]
> Also, note that, unfortunately, @arekbulski stopped working on Construct: #377 (comment), so most likely Construct target will be phased out in some time :(

You make it sound as if I removed Construct from the ecosystem. I said that I stopped working on it ie. I stopped adding new features. All existing features (and there is a lot of them) are still fully functional. Please do not drop support for Construct. Few days ago I resumed work on it. Currently there are 0 outstanding tickets. And the KSY export tool is still there too.

Looking at the CI page: I think you can remove "construct2" because I dropped support for 2.7 long time ago. Currently Construct runs only on 3.6+.

**Comment by Mimickal** [comment]
I may have found a bug in the Construct exporter. I have a type that depends on an enum, but the type is output in the Python before the enum. Since Python reads top-to-bottom, it raises a `NameError` about an undefined symbol.

Here is a minimal example:
```yaml
# The input ksy
meta:
  id: example
  endian: le
seq:
  - id: video
    type: video_settings
types:
  video_settings:
    seq:
      - id: texture_quality
        type: u1
        enum: quality_options
enums:
  quality_options:
    0: low
    1: medium
    2: high
```
```py
# The output Python
from construct import *
from construct.lib import *

example__video_settings = Struct(
	'texture_quality' / example__quality_options(Int8ub),
)

# This is required above!
def example__quality_options(subcon):
	return Enum(subcon,
		low=0,
		medium=1,
		high=2,
	)

example = Struct(
	'video' / LazyBound(lambda: example__video_settings),
)

_schema = example
```
Digging into the exporter code, it looks like it just always outputs types before enums https://github.com/kaitai-io/kaitai_struct_compiler/blob/0acfa60452a10b770a2c2e746f3de057f8ab76d9/shared/src/main/scala/io/kaitai/struct/ConstructClassCompiler.scala?ts=4#L36
Would reordering these two calls break anything else?

**Comment by Mimickal** [comment]
How do you use parameterized types with Construct? Are they currently implemented?

I have this ksy that represents a range with arbitrary numeric type.
```yaml
meta:
  id: range
  endian: le
params:
  - id: type_key
    type: str
seq:
  - id: min
    type: &type
      switch-on: type_key
      cases:
        '"f4"': f4
  - id: max
    type: *type
```
I use it like this:
```yaml
meta:
  id: wind
  endian: le
  imports:
    - common/range
seq:
  - id: velocity
    type: range("f4")
```

I expect Wind to give Range that "f4" value, and indeed in Python it does:
```python
# Stripped for brevity
# range.py
class Range(KaitaiStruct):
    def __init__(self, type_key, _io, _parent=None, _root=None):
        self.type_key = type_key

	def _read(self):
        _on = self.type_key
        if _on == u"f4":
			self.min = self._io.read_f4le()

# wind.py
from structs import range
class Wind(KaitaiStruct):
def _read(self):
        self.velocity = range.Range(u"f4", self._io)
```

In Construct, Range does appear to switch on `type_key`, but Wind does not seem to give Range the expected "f4" `type_key`. (Note, the import statement comes from my PR https://github.com/kaitai-io/kaitai_struct_compiler/pull/242)
```python
# range.py
range = Struct(
	'min' / Switch(this.type_key, {u"f4": Float32l, }),
	'max' / Switch(this.type_key, {u"f4": Float32l, }),
)

# wind.py
from structs.range import * # Added in my PR

wind = Struct(
	'velocity' / LazyBound(lambda: range), # Where is "f4"?
)
```

When attempting to load Wind data, I get `KeyError: 'type_key'`. I'm not quite sure where you would even add the expected `type_key` in Wind. According to the Construct docs, `LazyBound` does not type a parameter.
